index.js

```diff
@@ -64,82 +83,151 @@ function withFilter(predicate) {
+  var recordMap = {};
+
+  records.forEach(function(RecordType) {
+    var rec = new RecordType({});
+
+    if (!recordName(rec)) {
+      throw new Error('Cannot (de)serialize Record() without a name field');
+    }
+
+    recordMap[recordName(rec)] = RecordType;
+  });
+
+  var recordWriter = createWriter(predicate, recordMap);
+  var recordReader = createReader(recordMap);
+
+  return {
+    toJSON: function(data) {
+      return recordWriter.write(data);
+    },
+    fromJSON: function(data) {
+      return recordReader.read(data);
+    }
+  };
+}
+exports.withRecords = withRecords;
+
+function createWriter(predicate, recordMap) {
+  var handlers = transit.map([
+    Immutable.Map, transit.makeWriteHandler({
+      tag: function() {
+        return 'iM';
+      },
+      rep: function(m) {
+        var i = 0, a = new Array(2 * m.size);
+        if (predicate) {
+          m = m.filter(predicate);
+        }
+        m.forEach(function(v, k) {
+          a[i++] = k;
+          a[i++] = v;
+        });
+        return a;
+      }
+    }),
+    Immutable.OrderedMap, transit.makeWriteHandler({
+      tag: function() {
+        return 'iOM';
+      },
+      rep: function(m) {
+        var i = 0, a = new Array(2 * m.size);
+        if (predicate) {
+          m = m.filter(predicate);
+        }
+        m.forEach(function(v, k) {
+          a[i++] = k;
+          a[i++] = v;
+        });
+        return a;
+      }
+    }),
+    Immutable.List, transit.makeWriteHandler({
+      tag: function() {
+        return "iL";
+      },
+      rep: function(v) {
+        if (predicate) {
+          v = v.filter(predicate);
+        }
+        return v.toArray();
+      }
+    }),
+    Immutable.Set, transit.makeWriteHandler({
+      tag: function() {
+        return "iS";
+      },
+      rep: function(v) {
+        if (predicate) {
+          v = v.filter(predicate);
+        }
+        return v.toArray();
+      }
+    }),
+    Immutable.OrderedSet, transit.makeWriteHandler({
+      tag: function() {
+        return "iOS";
+      },
+      rep: function(v) {
+        if (predicate) {
+          v = v.filter(predicate);
+        }
+        return v.toArray();
+      }
+    }),
+    Function, transit.makeWriteHandler({
+      tag: function() {
+        return '_';
+      },
+      rep: function() {
+        return null;
+      }
+    }),
+    "default", transit.makeWriteHandler({
+      tag: function() {
+        return 'iM';
+      },
+      rep: function(m) {
+        if (!('toMap' in m)) {
+          var e = "Error serializing unrecognized object " + m.toString();
+          throw new Error(e);
+        }
+        m = m.toMap();
+
+        var i = 0, a = new Array(2 * m.size);
+        if (predicate) {
+          m = m.filter(predicate);
+        }
+        m.forEach(function(v, k) {
+          a[i++] = k;
+          a[i++] = v;
+        });
+        return a;
+      }
+    }),
+  ]);
+
+  for (var name in recordMap) {
```

Comment:
Would prefer `Object.keys(recordMap).forEach` here. I'm always a bit uneasy about `for..in` loops for some reason - probably an old habit.